### PR TITLE
Fixed Go To Definition using jsconfig

### LINKED ISSUE #46854: "Go To Definition" unnecessarily shows import definition when using jsconfig.json
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.62.1
- OS Version: Windows 11 Home (21H2)

Steps to Reproduce:

1. Clone [this](https://github.com/sefranz/vs-nav-test) minimal example
2. Open file `/src/pages/test.jsx`
3. Focus the jsx-tag `<HelloWorlder />`
4. Invoke "Go To Definition" (Ctrl+Click / via context menu / F12)
5. The focus jumps to the import statement in line 2

This behavior might be correct, but highly useless. This way, you have to always manually focus the imported filename and invoke the command again in order to open the corresponding file.

Also, this behavior does not occur when deleting the `jsconfig.json` file.

nav-test

### PR DESCRIPTION
**Pull request by armanio123**
Fixes #46854

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -3467,8 +3467,14 @@ namespace ts {
             const mode = contextSpecifier && isStringLiteralLike(contextSpecifier) ? getModeForUsageLocation(currentSourceFile, contextSpecifier) : currentSourceFile.impliedNodeFormat;
             const resolvedModule = getResolvedModule(currentSourceFile, moduleReference, mode);
             const resolutionDiagnostic = resolvedModule && getResolutionDiagnostic(compilerOptions, resolvedModule);
-            const sourceFile = resolvedModule && !resolutionDiagnostic && host.getSourceFile(resolvedModule.resolvedFileName);
+            const sourceFile = resolvedModule
+                && (!resolutionDiagnostic || resolutionDiagnostic === Diagnostics.Module_0_was_resolved_to_1_but_jsx_is_not_set)
+                && host.getSourceFile(resolvedModule.resolvedFileName);
             if (sourceFile) {
+                // If there's a resolutionDiagnostic we need to report it even if a sourceFile is found.
+                if(resolutionDiagnostic) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:3475]:
```suggestion
                if (resolutionDiagnostic) {
```

File: tests/cases/fourslash/goToDefinitionJsxNotSet.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//// const /*def*/Foo = () => (
+////     <div>foo</div>
+//// );
+//// export default Foo;
+
+// @Filename: /bar.jsx
+//// import Foo from './foo';
+//// const a = <[|/*use*/Foo|] />
+
+verify.goToDefinition("use", "def");
```
Comment (gabritto) [inline tests/cases/fourslash/goToDefinitionJsxNotSet.ts:17]: Is this test enough to reproduce the issue, even though there's no jsconfig in this test? From the issue, I thought the bug only happened if we had a jsconfig.

Comment (gabritto) [inline tests/cases/fourslash/goToDefinitionJsxNotSet.ts:17]: Is this test enough to reproduce the issue, even though there's no jsconfig in this test? From the issue, I thought the bug only happened if we had a jsconfig.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -3467,8 +3467,14 @@ namespace ts {
             const mode = contextSpecifier && isStringLiteralLike(contextSpecifier) ? getModeForUsageLocation(currentSourceFile, contextSpecifier) : currentSourceFile.impliedNodeFormat;
             const resolvedModule = getResolvedModule(currentSourceFile, moduleReference, mode);
             const resolutionDiagnostic = resolvedModule && getResolutionDiagnostic(compilerOptions, resolvedModule);
-            const sourceFile = resolvedModule && !resolutionDiagnostic && host.getSourceFile(resolvedModule.resolvedFileName);
+            const sourceFile = resolvedModule
+                && (!resolutionDiagnostic || resolutionDiagnostic === Diagnostics.Module_0_was_resolved_to_1_but_jsx_is_not_set)
```
Comment (gabritto) [inline src/compiler/checker.ts:3471]: just out of curiosity, how was that function affecting the behavior of `go to definition`? was the issue that `go to definition` couldn't get a resolved source file?

File: tests/cases/fourslash/goToDefinitionJsxNotSet.ts (hunk 2 above)
Comment (armanio123) [reply tests/cases/fourslash/goToDefinitionJsxNotSet.ts:17]: It's enough. I believe the use of `allowJs` validates the `jsconfig` scenario.

File: src/compiler/checker.ts (hunk 3 above)
Comment (armanio123) [reply src/compiler/checker.ts:3471]: That's correct. go to definition was not getting the resolved file because of `sourceFile` being `null`.

**Review by gabritto — APPROVED** [review]

File: src/compiler/checker.ts (hunk 1 above)
Comment (armanio123) [reply src/compiler/checker.ts:3475]: Done.

**Pull request merged**